pkg/controllers/management/oidcprovider/controller_test.go

```diff
@@ -0,0 +1,301 @@
+					},
+				}, nil)
+				p.secretClient.EXPECT().Update(&v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fakeClientId,
+						Namespace: secretNamespace,
+					},
+					Data: map[string][]byte{
+						secretKeyPrefix + "1": []byte(fakeClientSecret),
+					},
+				}).Return(nil, nil)
+				p.oidcClient.EXPECT().Update(&v3.OIDCClient{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:        fakeOIDCClientName,
+						Annotations: map[string]string{},
+					},
+					Status: v3.OIDCClientStatus{
+						ClientID: fakeClientId,
+					},
+				}).Return(nil, nil)
+			},
+		},
+		"client secret is removed with annotation": {
+			oidcClient: &v3.OIDCClient{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fakeOIDCClientName,
+					Annotations: map[string]string{
+						removeClientSecretAnn: secretKeyPrefix + "1",
+					},
+				},
+				Status: v3.OIDCClientStatus{
+					ClientID: fakeClientId,
+				},
+			},
+			setupMock: func(p *mockParams) {
+				p.secretCache.EXPECT().Get(secretNamespace, fakeClientId).Return(&v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fakeClientId,
+						Namespace: secretNamespace,
+					},
+					Data: map[string][]byte{
+						secretKeyPrefix + "1": []byte(fakeClientSecret),
+						secretKeyPrefix + "2": []byte(fakeClientSecret),
+					},
+				}, nil)
+				p.secretClient.EXPECT().Update(&v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fakeClientId,
+						Namespace: secretNamespace,
+					},
+					Data: map[string][]byte{
+						secretKeyPrefix + "2": []byte(fakeClientSecret),
+					},
+				}).Return(nil, nil)
+				p.oidcClient.EXPECT().Update(&v3.OIDCClient{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:        fakeOIDCClientName,
+						Annotations: map[string]string{},
+					},
+					Status: v3.OIDCClientStatus{
+						ClientID: fakeClientId,
+					},
+				}).Return(nil, nil)
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			mocks := &mockParams{
+				secretCache:     fake.NewMockCacheInterface[*v1.Secret](ctlr),
+				secretClient:    fake.NewMockClientInterface[*v1.Secret, *v1.SecretList](ctlr),
+				oidcClientCache: fake.NewMockNonNamespacedCacheInterface[*v3.OIDCClient](ctlr),
+				oidcClient:      fake.NewMockNonNamespacedClientInterface[*v3.OIDCClient, *v3.OIDCClientList](ctlr),
+				generator:       mocks.NewMockClientIDAndSecretGenerator(ctlr),
+			}
+			if test.setupMock != nil {
+				test.setupMock(mocks)
+			}
+
+			c := oidcClientController{
+				secretClient:    mocks.secretClient,
+				secretCache:     mocks.secretCache,
+				oidcClient:      mocks.oidcClient,
+				oidcClientCache: mocks.oidcClientCache,
+				generator:       mocks.generator,
+			}
+
+			_, err := c.onChange("", test.oidcClient)
+
+			if test.expectedErr != "" {
+				assert.ErrorContains(t, err, test.expectedErr)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestOnRemove(t *testing.T) {
```

Comment:
If you set up the ownership you don't need this 😄 

Just assert that the `Secret` is owned by the `OIDCClient`.